src/coreclr/src/inc/clrconfigvalues.h

```diff
@@ -395,6 +395,8 @@ RETAIL_CONFIG_DWORD_INFO(INTERNAL_TraceInterpreterIL, W("TraceInterpreterIL"), 0
 RETAIL_CONFIG_DWORD_INFO(INTERNAL_TraceInterpreterOstack, W("TraceInterpreterOstack"), 0, "Logs operand stack after each IL instruction of interpreted methods to the console")
 CONFIG_DWORD_INFO(INTERNAL_TraceInterpreterVerbose, W("TraceInterpreterVerbose"), 0, "Logs interpreter progress with detailed messages to the console")
 CONFIG_DWORD_INFO(INTERNAL_TraceInterpreterJITTransition, W("TraceInterpreterJITTransition"), 0, "Logs when the interpreter determines a method should be JITted")
+RETAIL_CONFIG_DWORD_INFO(INTERNAL_ForceJitOptimization, W("ForceJitOptimization"), 0, "If non-zero, force JIT optimization to be enabled") // this has no effect in Release
```

Comment:
I'll have to look farther in this PR, but this name (ForceJitOptimization) doesn't seem appropriate.

Also, would it be useful to have all interpreter-related COMPlus variable names use `Interpreter` as a prefix (as most or all currently do)?

Reply:
Hi @BruceForstall, thanks for pointing that out! The name `ForceJitOptimization` seems to describe what the flag does in the following place:
https://github.com/dotnet/runtime/blob/f01109b9c5e4402d991e391206ccd30d1efedcee/src/coreclr/src/vm/method.cpp#L4868
I feel that it is similar to @noahfalk's suggestion `AllowOptimizationForDebugCode`.

I agree that it would be best if the interpreter-related names have `Interpreter` as a prefix. Do you have recommendations for these names?

Reply:
The current implementation isn't forcing that jit optimization must occur, rather it permits optimized compilation to possibly occur. If all the other criteria are met the method will be marked as being eligible for tiered compilation, and then if the method is executed frequently enough the tier1 version will be compiled that contains optimizations. I'd suggest breaking the discussion in two parts:
1. Is the implemented behavior useful or do we actually want something else?
2. If that is the behavior we want, what is a good name for it?

Thinking on it a bit I can't tell if this is useful behavior? If the goal is to test performance I assume we'd want to do that on a release build. If the goal is to test correctness of the interpreter I assume we'd want to use the interpreter as much as possible. So far the only scenario I've come up with where we'd want to tier up from interpreter code to jitted code in Debug is if we need to run lots of test code that is already Debug configuration but the performance is so bad that we have to tier up the frequently executed code paths to let the tests run in a reasonable timeframe. Of course there could be another use you had in mind that I am not thinking of?

Reply:
Thanks for the comment, @noahfalk ! My problem with the name it is sounds like it is forcing the JIT to generate code with certain characteristics, which is not the point here at all (and doesn't affect what the JIT will do). But I agree that it makes sense to better understand the scenario here.

Reply:
Hi @noahfalk, thanks for your reply! I think it is a good point to think about the use scenarios of the added `COMPlus` variables. I agree that the two scenarios you came up with (i.e. testing performance in Release` and testing functionality with interpreter triggered as much as possible) are the two main scenarios in my mind. My initial thought to trigger the interpreter only in Tier 0 in Debug/checked is due to the following two scenarios:

1. To mimic the behavior of the Runtime in Release so we can also test the correctness of promotion from Tier 0 to Tier 1.
2. In the future when CI is set up for the interpreter, to check the assertions in CI for the interpreter in the Checked configuration.

Do you think that these two are good scenarios to consider?

Reply:
I renamed `ForceJitOptimization` to `AllowJitOptimization` in commit https://github.com/dotnet/runtime/pull/37825/commits/7cdd55f830ca869aeee386ee4945d41ed588f7b9.

Reply:
I want to clarify something here: your comment in the code mentions that this only applies to Debug/Checked code and not Release; are you referring to the runtime or the C# application?  My understanding of this flag was that it was similar to enabling JIT optimizations when debugging a managed application in VS, i.e., that you are allowing the JIT to optimize *debug IL code*.  Does the runtime's configuration have any impact on this flag?

Reply:
Hi @josalem! I think the description is actually confusing here. What I mean by "(for Debug/Checked)" is that this variable is only effective for Debug/Checked, but it does not have an effect for Release (I'm referring to the C# application). The runtime's configuration does not have impact on this. This variable itself does not check the configuration of the application; I'm just talking about the effects it has.

However I agree that this is not a good name nor a good description. Do you have some recommendation for this?

Reply:
> Debug/Checked

Checked means optimizations on + assertions on. Checked is like Release as far as managed optimizations are concerned: https://github.com/dotnet/runtime/blob/master/src/coreclr/src/System.Private.CoreLib/System.Private.CoreLib.csproj#L87 . 

Checked is special flavor for used for CoreCLR itself only. Libraries or applications are just Debug/Release.

Reply:
> Do you think that these two are good scenarios to consider?
1. To mimic the behavior of the Runtime in Release so we can also test the correctness of promotion from Tier 0 to Tier 1.

Did you mean mimic the behavior of a _managed assembly_ in Release? Our CI builds already compile the runtime in Release. I think the most straightforward way right now to test that tier0->tier1 promotion occurs is to use a managed assembly that was compiled in Release configuration. 

> 2. In the future when CI is set up for the interpreter, to check the assertions in CI for the interpreter in the Checked configuration.

If you are refering to assertions in the runtime source code that solely depends on whether the runtime has been compiled in Checked configuration. The config var would be unneeded there. If you are refering to Debug.Assert() calls in the C# source code of test apps then the "UseInterpretterAlways" config appears to handle that case already.

If these suggestions work then it appears there would be no need for the config switch that modifies JitOptimization policies, and then no need to worry about what to name it ; )